floodpipeline/data.py

```diff
@@ -0,0 +1,515 @@
+        return any([x.triggered for x in self.data_units])
+
+
+class BasinDataSet:
+    """Base class for basin data sets"""
+
+    def __init__(
+        self,
+        country: str = None,
+        timestamp: datetime = datetime.now(),
+        data_units: List[BasinDataUnit] = None,
+    ):
+        self.country = country
+        self.timestamp = timestamp
+        self.data_units = data_units or []
+
+    def get_pcodes(self, adm_level: int = None):
+        """Return list of unique pcodes, optionally filtered by adm_level"""
+        return list(set([x.pcodes for x in self.data_units]))
+    
+
+
+    def get_basin_codes(self):
+        """Return list of unique station codes"""
+        return list(
+            set([x.hybasid for x in self.data_units if hasattr(x, "hybasid")])
+        )     
+
+    def get_lead_times(self):
+        """Return list of unique lead times"""
+        return list(
+            set([x.lead_time for x in self.data_units if hasattr(x, "lead_time")])
+        )
+
+    def get_data_unit(self, hybasid: str, lead_time: float = None) -> BasinDataUnit:
+        if lead_time is not None:
+            bdu = next((x for x in self.data_units if x.hybasid == hybasid and x.lead_time == lead_time), None)
+        else:
+            bdu = next((x for x in self.data_units if x.hybasid == hybasid), None)
+        if not bdu:
+            raise ValueError(f"Data unit with basin_id {hybasid} not found")
+        return bdu
+
+    def upsert_data_unit(self, data_unit: BasinDataUnit):
+        existing = next((i for i,x in enumerate(self.data_units)
+                         if x.hybasid == data_unit.hybasid and getattr(x, 'lead_time', None) == getattr(data_unit, 'lead_time', None)), None)
+        if existing is None:
+            self.data_units.append(data_unit)
+        else:
+            self.data_units[existing] = data_unit
+
+class StationDataSet:
+    """Base class for station data sets"""
+
+    def __init__(
+        self,
+        country: str = None,
+        timestamp: datetime = datetime.now(),
+        data_units: List[StationDataUnit] = None,
+    ):
+        self.country = country
+        self.timestamp = timestamp
+        self.data_units = data_units
+
+    def get_data_unit(
+        self, station_code: str, lead_time: int = None
+    ) -> StationDataUnit:
+        """Get data unit by station_code and optionally by lead time"""
+        if not self.data_units:
+            raise ValueError("Data units not found")
+        if lead_time:
+            bdu = next(
+                filter(
+                    lambda x: x.station_code == station_code
+                    and x.lead_time == lead_time,
+                    self.data_units,
+                ),
+                None,
+            )
+        else:
+            bdu = next(
+                filter(lambda x: x.station_code == station_code, self.data_units),
+                None,
+            )
+        if not bdu:
+            raise ValueError(
```

Comment:
Does raising this error end the pipeline run? Would it be possible to skip this step on error and continue with the pipeline run? Then it will be possible to see a partial forecast instead of no forecast.